### ISSUE TITLE
[RMP] Provide PyTorch serving support for T4R models in Torchscript

### ISSUE THREAD

**Issue by karlhigley** [NVIDIA-Merlin/Merlin#255]
## Problem:

Users should be able to serve pytorch models that were produced with Transformers4Rec or any other process using a Systems ensemble. This will work towards supporting Session-based models as well as expanding System's support for a new modeling framework.

## Goal:

Systems should be able to serve all pytorch models that are currently supported by Triton.

## Definition of Done
Have an example that serves PyT session based model in conjunction with a NVT workflow where the session based models scores the whole catalog

## Constraints:

Not all pytorch models can be served via Triton's `pytorch` backend, so we will need to be able to use multiple backends in order to serve all Triton-compatible pytorch models.

### Transformers4Rec
- [x] Update T4R to use Merlin std schema
  - [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/pull/507
- [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/499
- [x] NVIDIA-Merlin/systems#322
- [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/668

### Systems

- [x] Use Triton's `pytorch` backend for "torchscriptable" models for optimized performance
  - [x] NVIDIA-Merlin/systems#153
  - [x] https://github.com/NVIDIA-Merlin/systems/pull/176
- [x] Load the schema files written out by T4R 

### Integration Issues

- [ ] Address [the issues](https://docs.google.com/document/d/1Wp_G_Bifrok0KvQUgQ2FP40oOSI3lvGqDH48YaK4IwI/edit?usp=sharing) @rnyak surfaced while trying to put together an example
  - [x] Fix missing value_count property and/or error from value_count op if sequence min and max length is same: This  PR https://github.com/NVIDIA-Merlin/core/pull/173 should fix that issue.
  - [x] Fix wrong response dimension from TIS after sending request. This PR https://github.com/NVIDIA-Merlin/Transformers4Rec/pull/543 should be fixing that issue.
  - [x] https://github.com/NVIDIA-Merlin/systems/issues/234
  - [x] https://github.com/NVIDIA-Merlin/core/issues/175
  - [x] https://github.com/NVIDIA-Merlin/systems/issues/240

Nice to have: (P1)
- [x] https://github.com/NVIDIA-Merlin/systems/issues/135

### Documentation
- [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/491
- [x] https://github.com/NVIDIA-Merlin/systems/issues/257
- [x] NVIDIA-Merlin/Merlin#904
- [x] NVIDIA-Merlin/Transformers4Rec#692
- [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/664

### Examples
- [x] NVIDIA-Merlin/Transformers4Rec#660
- [x] https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/676

### Blockers:
- [ ] [[INF] Unresolved architectural decisions](https://github.com/NVIDIA-Merlin/Merlin/issues/776)
- [x] Support for ragged tensors in T4R
- [x] Start with fix padding and have it all padded to the same length and then investigate whether it's worthwhile to add the like padding to the Max Sequence length
- [x] Padding support in dataloader that supports systems along with cross framework support

**Comment by viswa-nvidia** [comment]
@karlhigley , I saw the comment after 22.05 and wondered whether we should have this ticket defined and ready for discussion at the next grooming meeting. Can you help in definition ?

**Comment by viswa-nvidia** [comment]
@karl, do you see this getting over by 22.08?

**Comment by viswa-nvidia** [comment]
@karlhigley , should this ticket also have these additional requirements that we discussed in the backlog grooming meeting?

- enable PyT T4R 
     - Multi-gpu training
             - DP (working)
     - Serving
             - Python-backend (working)

**Comment by karlhigley** [comment]
No, because they're not part of the scope as described at the top of this issue and they've already been completed

**Comment by viswa-nvidia** [comment]
@karlhigley , the original ask from product team  is Session based recommender - Add PyT API to Merlin models / Systems. This ticket only covers  the systems side and not models. I think this should be a [Task] that must be part of a RMP ticket that will cover the models side also. Let me know your thoughts on this.

**Comment by karlhigley** [comment]
@viswa-nvidia That was my original thought too, until I got talked out of it. 🤷🏻

**Comment by EvenOldridge** [comment]
@edknv does the work you've done for Transformers4Rec support this functionality from the model training side?
@karlhigley excluding onnx support can we currently serve Torchscript based models in systems?
Looking to close this

**Comment by karlhigley** [comment]
Each side of this equation works independently, but no one has tested them together AFAIK, so we're not ready to close this yet

**Comment by EvenOldridge** [comment]
@rnyak @bschifferer can one of you please test this functionality and ensure that we're able to save models via torchscript and serve via systems.  I think this is relevant to a number of your customers.

**Comment by karlhigley** [comment]
I'm working on an integration test for that here, but it isn't working yet: https://github.com/NVIDIA-Merlin/systems/pull/176

**Comment by viswa-nvidia** [comment]
@marcromeyn , Add tickets to the backlog for the optimizations
